src/main/java/Board.java

```diff
@@ -39,43 +69,72 @@ public void clearBoard() {
         }
     }
 
-    @NotNull
-    public List<Cell> getAdjacentCells(@NotNull Position cellPosition) throws InvalidPositionException {
-        if (cellPosition.getRow() > numberOfRows || cellPosition.getColumn() > numberOfColumns) {
+    public Set<Position> getAdjacentPositions(Position position) throws InvalidPositionException {
+        if (isPositionOutOfBoardBounds(position)) {
             throw new InvalidPositionException("The specified position is outside the board");
         }
-        List<Cell> adjacentCells = new ArrayList<>();
+        Set<Position> adjacentPositions = new HashSet<>(8);
         for (int i = -1; i <= 1; i++) {
             for (int j = -1; j <= 1; j++) {
                 try {
-                    if (j == 0 && i == 0) continue;
-                    Position adjacentCellPosition = new Position(cellPosition.getRow() + i, cellPosition.getColumn() + j);
-                    adjacentCells.add(cells.get(adjacentCellPosition));
+                    if (i == 0 && j == 0) continue;
+                    adjacentPositions.add(new Position(position.getRow() + i, position.getColumn() + j));
                 } catch (InvalidPositionException ignored) {
-                    // if there are no adjacent cells in one direction, simply don't add them to the list
                 }
             }
         }
-        return adjacentCells;
+        return adjacentPositions;
+    }
+
+    private boolean isPositionOutOfBoardBounds(Position position) {
+        return position.getRow() > numberOfRows || position.getColumn() > numberOfColumns;
     }
 
     public boolean areAdjacentCellsOccupied(@NotNull Position cellPosition) throws InvalidPositionException {
-        List<Cell> adjacentCell = getAdjacentCells(cellPosition);
-        for(Cell cell : adjacentCell) {
-            if(!cell.isOccupied()) return false;
+        Set<Position> adjacentPositions = getAdjacentPositions(cellPosition);
+        for (Position position : adjacentPositions) {
+            if (!isCellOccupied(position)) return false;
         }
         return true;
     }
 
     public boolean hasBoardMoreThanOneFreeCell() {
         int numberOfFreeCellsOnBoard = 0;
         for (Map.Entry<Position, Cell> cellOnBoard : cells.entrySet()) {
-            if (!cellOnBoard.getValue().isOccupied()) numberOfFreeCellsOnBoard++;
+            if (!isCellOccupied(cellOnBoard.getKey())) numberOfFreeCellsOnBoard++;
             if (numberOfFreeCellsOnBoard > 1) return true;
         }
         return false;
     }
 
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (int i = getNumberOfRows(); i > 0; i--) {
```

Comment:
Board fields can be used here

Reply:
Yes of course, I just copied and pasted and I forgot to change this